spring-kafka/src/main/java/org/springframework/kafka/config/RetryTopicConfigurationSupport.java

```diff
@@ -0,0 +1,485 @@
+	 * back off partitions.
+	 * To configure it, override the {@link #configureKafkaBackOffManager} method.
+	 * To provide a custom implementation, either override this method, or
+	 * override the {@link RetryTopicComponentFactory#kafkaBackOffManagerFactory} method
+	 * and return a different {@link KafkaBackOffManagerFactory}.
+	 * @param registry the {@link ListenerContainerRegistry} to be used to fetch the
+	 * {@link MessageListenerContainer} at runtime to be backed off.
+	 * @param taskExecutor the {@link TaskExecutor} to be used with the
+	 * {@link KafkaConsumerTimingAdjuster}.
+	 * @return the instance.
+	 */
+	@Bean(name = KafkaListenerConfigUtils.KAFKA_CONSUMER_BACK_OFF_MANAGER_BEAN_NAME)
+	public KafkaConsumerBackoffManager kafkaConsumerBackoffManager(
+			@Qualifier(KafkaListenerConfigUtils.KAFKA_LISTENER_ENDPOINT_REGISTRY_BEAN_NAME)
+					ListenerContainerRegistry registry,
+			@Qualifier(BACK_OFF_MANAGER_THREAD_EXECUTOR_BEAN_NAME) TaskExecutor taskExecutor) {
+
+		KafkaBackOffManagerFactory backOffManagerFactory =
+				this.componentFactory.kafkaBackOffManagerFactory(registry);
+		JavaUtils.INSTANCE.acceptIfInstanceOf(PartitionPausingBackOffManagerFactory.class, backOffManagerFactory,
+				factory -> configurePartitionPausingFactory(taskExecutor, factory));
+		return backOffManagerFactory.create();
+	}
+
+	/**
+	 * Internal method for processing the {@link PartitionPausingBackOffManagerFactory}.
+	 * @param taskExecutor the {@link TaskExecutor} instance to be used with
+	 * {@link WakingKafkaConsumerTimingAdjuster}. Consider overriding the
+	 * {@link #configureKafkaBackOffManager} method for furher customization.
+	 * @param factory the factory instance.
+	 */
+	private void configurePartitionPausingFactory(TaskExecutor taskExecutor,
+												PartitionPausingBackOffManagerFactory factory) {
+		KafkaBackOffManagerConfigurer configurer = new KafkaBackOffManagerConfigurer();
+		configureKafkaBackOffManager(configurer);
+		Assert.isTrue(!configurer.timingAdjustmentEnabled
+						|| configurer.maxThreadPoolSize == null
+						|| ThreadPoolTaskExecutor.class.isAssignableFrom(taskExecutor.getClass()),
+				() -> "TaskExecutor must be an instance of ThreadPoolTaskExecutor to set maxThreadPoolSize");
+		factory.setTimingAdjustmentEnabled(configurer.timingAdjustmentEnabled);
+		JavaUtils.INSTANCE
+				.acceptIfNotNull(configurer.maxThreadPoolSize, poolSize -> ((ThreadPoolTaskExecutor) taskExecutor)
+						.setMaxPoolSize(poolSize))
+				.acceptIfCondition(configurer.timingAdjustmentEnabled, taskExecutor, factory::setTaskExecutor)
+				.acceptIfNotNull(configurer.clock, factory::setClock);
+	}
+
+	/**
+	 * Create the {@link TaskExecutor} instance that will be used with the
+	 * {@link WakingKafkaConsumerTimingAdjuster}, if timing adjustment is enabled.
+	 * @return the instance.
+	 */
+	@Bean(name = BACK_OFF_MANAGER_THREAD_EXECUTOR_BEAN_NAME)
+	public TaskExecutor backoffManagerTaskExecutor() {
```

Comment:
Sure, and since this would probably be the only actual breaking change for this PR, that'd be a benefit of rolling this back too. At least we gave it a shot 😄 Thanks.